### PR TITLE
139 fix interval mapping

### LINKED ISSUE #139: Fix Function create_intervals_mapping in GetVaccinationData
The function calculates mappings from multiple intervals to different multiple intervals.
This works for the case that 1 interval is mapped onto two others, but not from 1 to more than 2.
This could be generalized.

For example:
`[0, 10, 100]`
onto
`[0, 50, 60, 70, 100]`
results in
```
0:[[1, 0]]
1:[[0.5555555555555556, 0], [0.6666666666666667, 1], [0.7777777777777779, 2], [0.2222222222222221, 3]]
```
but should be

```
0:[[1, 0]]
1:[[0.444, 0], [0.111, 1], [0.111, 2], [0.333, 3]]
```

### PR DESCRIPTION
**Pull request by annette-lutz**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [ ] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request.
* [ ] Coverage report for new code is acceptable. 

closes #139

### CODE REVIEW

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py [hunk 1]
```diff
@@ -60,32 +60,32 @@ def create_intervals_mapping(from_lower_bounds, to_lower_bounds):
     @param from_lower_bounds lower bounds of original intervals
     @param to_lower_bounds desired lower bounds of new intervals
     @return mapping from intervals to intervals
+        The mapping is given as a list containing for every original interval a
+        list of tupels of the share of the original interval that is mapped to
+        the new interval and the index of the new interval
```
[the remark this answers is not in the document]
Comment (xsaschako) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py:65]: Also: Maybe describe which cases work, e.g. the boundaries of both intervals have to be the same. E.g. [0,10,100]->[10,20,90] doesn't work. I think it has to be a->b a has to be included in b or something like that

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py [hunk 2]
```diff
@@ -60,32 +60,32 @@ def create_intervals_mapping(from_lower_bounds, to_lower_bounds):
...
+            j += 1
+        while from_lower_bounds[i+1] > to_lower_bounds[j+1]:
+            # compute share of interval i that is send to interval j
+            len_interval_i = from_lower_bounds[i+1] - from_lower_bounds[i]
+            share = (to_lower_bounds[j+1] - pos) / len_interval_i
+            from_to_mapping[i].append([share, j])
+            remaining_share -= share
+            pos = to_lower_bounds[j+1]
+            j += 1
+        # if upper bound of new interval j is not smaller than upper bound of
```
Comment (xsaschako) [inline pycode/memilio-epidata/memilio/epidata/getVaccinationData.py:86]: Looks clean and good.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py [hunk 3]
```diff
@@ -317,31 +318,35 @@ def test_intervall_mapping(self):
...
         for test_map, calculated_map in zip(test_map1, map_bounds1):
-            self.assertTrue(
-                np.allclose(
-                    np.array(test_map),
-                    np.array(calculated_map),
-                    rtol=1e-05),
-                "Not the same Arrays")
+            for test_val, calculated_val in zip(test_map, calculated_map):
+                self.assertEqual(test_val[1], calculated_val[1])
+                self.assertAlmostEqual(test_val[0], calculated_val[0])
```
Comment (xsaschako) [inline pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py:339]: That is indeed cleaner.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py [hunk 4]
```diff
@@ -293,14 +293,15 @@ def test_split_column_based_on_values_ident_length_wrong(self):
                     column_names_i[j],
                     'Column names do not match.')
 
-    @unittest.skip("Skip Test till create_intervals_mapping is fixed")
-    def test_intervall_mapping(self):
-        lower_bounds1 = np.array(
+    def test_interval_mapping(self):
```
Comment (xsaschako) [inline pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py:296]: Maybe add some tests for edge cases where the bounds of the intervalls don't match

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py [hunk 5]
```diff
@@ -60,32 +60,32 @@ def create_intervals_mapping(from_lower_bounds, to_lower_bounds):
...
-            from_to_mapping[i].append([1-share, j])
-            # if both upper bounds are equal, then all ages j will not get any more share from any old group
-            if from_lower_bounds[i+1] == to_lower_bounds[j+1]:
-                j += 1
-
+    j = 0  # iterator over new intervals
+    # iterate over original intervals
+    for i in range(0, len(from_lower_bounds) - 1):
+        remaining_share = 1  # share of original interval i to be distributed
+        # position up to which the distribution is already computed
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py:86]: Since the line is to long, pep8 changes the comment to be above the variable definition.

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py]: Changed the description.

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py [hunk 6]
```diff
@@ -60,32 +60,32 @@ def create_intervals_mapping(from_lower_bounds, to_lower_bounds):
...
-                j += 1
-            from_to_mapping[i].append([1-share, j])
-            # if both upper bounds are equal, then all ages j will not get any more share from any old group
-            if from_lower_bounds[i+1] == to_lower_bounds[j+1]:
-                j += 1
-
+    j = 0  # iterator over new intervals
+    # iterate over original intervals
+    for i in range(0, len(from_lower_bounds) - 1):
+        remaining_share = 1  # share of original interval i to be distributed
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py]: It specifies of which interval the remaining share is taken and makes it easier to understand which distribution is meant in the description of the position in the next line.

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py [hunk 7]
```diff
@@ -60,32 +60,32 @@ def create_intervals_mapping(from_lower_bounds, to_lower_bounds):
...
+        # position up to which the distribution is already computed
+        pos = from_lower_bounds[i]
+        # find first new interval j insecting original interval i
+        while from_lower_bounds[i] >= to_lower_bounds[j+1]:
+            j += 1
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py]: Since i and j are only indecies, I didn't call them intervals. Renamed them into from_idx and to_idx to easier distinguish between them.

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py (hunk 2 above)
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py]: Thanks.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py (hunk 3 above)
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py:345]: It helped me to debug my code :)

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py (hunk 4 above)
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_vaccination_data.py:296]: Added a test case with invalid boundaries and changed one to have non matching boundaries.

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py [hunk 8]
```diff
@@ -60,32 +59,47 @@ def create_intervals_mapping(from_lower_bounds, to_lower_bounds):
...
+        that is mapped to the new interval and the index of the new interval. We
+        assume that the range of the intervals mapped from is contained in the 
+        range of the intervals mapped to.
+        For example for from_lower_bounds = [5,20,30,80,85,90] and 
+        to_lower_bounds=[0,15,20,60,100] given the mapping would be
+        [[[2/3,0], [1/3,1]],
+         [[1,2]],
+         [[3/5,2], [2/5,3]],
+         [[1,3]],
+         [[1,3]]]
```
Comment (xsaschako) [inline pycode/memilio-epidata/memilio/epidata/getVaccinationData.py:74]: nice!

File: pycode/memilio-epidata/memilio/epidata/getVaccinationData.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (xsaschako) [reply pycode/memilio-epidata/memilio/epidata/getVaccinationData.py:86]: Okay then nevermind.

**Review by xsaschako — DISMISSED** [review]

**Review by dabele — APPROVED** [review]
No complaints.

Just a note, such an interpolation (or resampling) can be nicely represented as a matrix vector product. 

For example the matrix
```
[1, 0.5,   0,   0,   0,
 0, 0.5, 0.3,  0,   0,
 0,    0, 0,7,  1,   0,
 0,    0,    0,  0,   1]  
```
transforms [10, 10, 10, 10, 10] to [15, 8, 17, 10]. Maybe can be used in the future.

**Review by xsaschako — APPROVED** [review]
Approved!
